README.pod
=pod

=head1 NAME

Mojolicious::Plugin::TemplateToolkit - Template Toolkit renderer plugin for
Mojolicious

=head1 SYNOPSIS

 # Mojolicious
 $app->plugin('TemplateToolkit');
 $app->plugin(TemplateToolkit => {name => 'foo'});
 $app->plugin(TemplateToolkit => {template => {INTERPOLATE => 1}});
 
 # Mojolicious::Lite
 plugin 'TemplateToolkit';
 plugin TemplateToolkit => {name => 'foo'};
 plugin TemplateToolkit => {template => {INTERPOLATE => 1}});
 
 # Set as default handler
 $app->renderer->default_handler('tt2');
 
 # Render without setting as default handler
 $c->render(template => 'bar', handler => 'tt2');

=head1 DESCRIPTION

L<Mojolicious::Plugin::TemplateToolkit> is a renderer for C<tt2> or
C<Template Toolkit> templates. See L<Template> and L<Template::Manual> for
details on the C<Template Toolkit> format, and L<Mojolicious::Guides::Rendering>
for general information on rendering in L<Mojolicious>.

Along with template files, inline and data section templates can be rendered in
the standard Mojolicious fashion. Template files and data sections will be
retrieved using L<Mojolicious::Renderer> via L<Template::Provider::Mojo> for
both direct rendering and directives such as C<INCLUDE>. This means that
instead of specifying L<INCLUDE_PATH|Template::Manual::Config/"INCLUDE_PATH">,
you should set L<Mojolicious::Renderer/"paths"> to the appropriate paths.

 $app->renderer->paths(['/path/to/templates']);
 push @{$app->renderer->paths}, '/path/to/more/templates';

L<Mojolicious> stash values will be exposed directly as
L<variables|Template::Manual::Variables> in the templates, and the current
controller object will be available as C<c> or C<self>, similar to
L<Mojolicious::Plugin::EPRenderer>.

Helper methods can be called on the controller object as normal, as well as on
a L<proxy object|Mojolicious::Controller/"helpers"> available as C<h>. The
proxy object was previously needed for efficiency as the controller object used
C<AUTOLOAD> to call helper methods, but since Mojolicious 8.04 the controller
object uses a more efficient mechanism and the proxy object is no longer
needed. See L<Mojolicious::Plugin::DefaultHelpers> and
L<Mojolicious::Plugin::TagHelpers> for a list of all built-in helper methods.

Accessing helper methods directly as variables, rather than via the controller
or proxy object, is deprecated and may be removed in a future release.

 $c->stash(description => 'template engine');
 $c->stash(engines => [qw(Template::Toolkit Text::Template)]);
 
 [% FOREACH engine IN engines %]
   [% engine %] is a [% description %].
 [% END %]
 
 [% c.link_to('Template Toolkit', 'http://www.template-toolkit.org') %]
 
 [% c.param('foo') %]


=head1 OPTIONS

L<Mojolicious::Plugin::TemplateToolkit> supports the following options.

=head2 name

 # Mojolicious::Lite
 plugin TemplateToolkit => {name => 'foo'};

Handler name, defaults to C<tt2>.

=head2 template

 # Mojolicious::Lite
 plugin TemplateToolkit => {template => {INTERPOLATE => 1}};

Configuration values passed to L<Template> object used to render templates.
Note that L<Template::Provider::Mojo> will use L<Mojolicious::Renderer/"paths">
to find templates, not L<INCLUDE_PATH|Template::Manual::Config/"INCLUDE_PATH">
specified here.

=head1 METHODS

L<Mojolicious::Plugin::TemplateToolkit> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

 $plugin->register(Mojolicious->new);
 $plugin->register(Mojolicious->new, {name => 'foo'});

Register renderer in L<Mojolicious> application.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojolicious::Renderer>, L<Template>, L<Template::Provider::Mojo>

=cut


Changes
{{$NEXT}}

0.006     2021-06-16 01:04:48 EDT
  - Update IRC metadata to libera.chat

0.005     2018-10-18 19:16:25 EDT
  - Added a proxy object for calling helpers more efficiently as 'h'

0.004     2018-10-16 19:32:20 EDT
  - Deprecated calling helpers directly as template variables in favor of
    calling them via the 'c' controller variable, to disambiguate helpers from
    stash variables

0.003     2018-06-14 17:45:49 EDT
  - Removed randomly failing OUTLINE_TAG test
  - Fix rendering and including empty data templates

0.002     2015-10-20 20:03:54 EDT
  - Added Template::Provider::Mojo to resolve templates using
    Mojolicious::Renderer directly

0.001     2015-10-20 01:43:29 EDT
  - First release


Build.PL
# This Build.PL for Mojolicious-Plugin-TemplateToolkit was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojolicious-Plugin-TemplateToolkit was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.012.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "Template Toolkit renderer plugin for Mojolicious",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojolicious-Plugin-TemplateToolkit",
      "dist_version" => "0.004",
      "license" => "artistic_2",
      "module_name" => "Mojolicious::Plugin::TemplateToolkit",
      "recursive_test_files" => 1,
      "requires" => {
        "Class::Method::Modifiers" => 0,
        "Mojolicious" => "6.51",
        "Scalar::Util" => 0,
        "Template" => "2.26",
        "Template::Provider" => "2.94",
        "parent" => 0,
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88"
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}


lib/Template/Provider/Mojo.pm
package Template::Provider::Mojo;

use strict;
use warnings;
use parent 'Template::Provider';

use Class::Method::Modifiers ();
use Mojo::Util;
use Mojolicious::Renderer;
use Scalar::Util;
use Template::Constants;

our $VERSION = '0.007';

Class::Method::Modifiers::after '_init' => sub {
	my ($self, $params) = @_;
	Scalar::Util::weaken($self->{MOJO_RENDERER} = delete $params->{MOJO_RENDERER});
	$self->{ENCODING} //= $self->_mojo_renderer->encoding;
};

sub fetch {
	my ($self, $name) = @_;
	my ($data, $error);
	
	if (ref $name) {
		# $name can be a reference to a scalar, GLOB or file handle
		($data, $error) = $self->_load($name);
		($data, $error) = $self->_compile($data) unless $error;
		$data = $data->{data} unless $error;
	} else {
		# Use renderer to find template
		my $renderer = $self->_mojo_renderer;
		my $options = _template_options($renderer, $name);
		
		# Try template
		if (defined(my $path = $renderer->template_path($options))) {
			($data, $error) = $self->_fetch($path);
		}
		
		# Try DATA section
		elsif (defined(my $d = $renderer->get_data_template($options))) {
			($data, $error) = $self->_load(\$d);
			($data, $error) = $self->_compile($data) unless $error;
			$data = $data->{data} unless $error;
		}
		
		# No template
		else { ($data, $error) = (undef, Template::Constants::STATUS_DECLINED) }
	}
	
	return ($data, $error);
}

sub load {
	my ($self, $name) = @_;
	my ($data, $error);
	
	# Use renderer to find template
	my $renderer = $self->_mojo_renderer;
	my $options = _template_options($renderer, $name);
	
	# Try template
	if (defined(my $path = $renderer->template_path($options))) {
		($data, $error) = $self->_template_content($path);
	}
	
	# Try DATA section
	elsif (defined(my $d = $renderer->get_data_template($options))) {
		# Content is expected to be encoded
		$d = Mojo::Util::encode $self->{ENCODING}, $d if $self->{UNICODE} and $self->{ENCODING};
		($data, $error) = ($d, undef);
	}
	
	# No template
	else { return (undef, Template::Constants::STATUS_DECLINED) }
	
	if ($error) {
		return $self->{TOLERANT} ? (undef, Template::Constants::STATUS_DECLINED)
			: ($error, Template::Constants::STATUS_ERROR);
	} else {
		return ($data, Template::Constants::STATUS_OK);
	}
}

sub _mojo_renderer { shift->{MOJO_RENDERER} //= Mojolicious::Renderer->new }

# Split template name back into options
sub _template_options {
	my ($renderer, $name) = @_;
	my $options = {};
	if ($name =~ m/^(.+)\.(.+)\.(.+)\z/) {
		$options->{template} = $1;
		$options->{format} = $2;
		$options->{handler} = $3;
	} elsif ($name =~ m/^(.+)\.(.+)\z/) {
		$options->{template} = $1;
		$options->{format} = $2;
	} else {
		$options->{template} = $name;
		$options->{format} = $renderer->default_format;
		$options->{handler} = $renderer->default_handler;
	}
	return $options;
}

1;

=head1 NAME

Template::Provider::Mojo - Use Mojolicious to provide templates

=head1 SYNOPSIS

 my $app = Mojolicious->new;
 $provider = Template::Provider::Mojo->new({MOJO_RENDERER => $app->renderer});
 
 ($template, $error) = $provider->fetch($name);

=head1 DESCRIPTION

L<Template::Provider::Mojo> is a L<Template::Provider> subclass that uses a
L<Mojolicious::Renderer> instance to resolve template names. This means that
L<Mojolicious::Renderer/"paths"> will be searched for file-based templates, and
L<Mojolicious::Renderer/"classes"> will be searched for DATA templates. The
C<ENCODING> configuration setting will be initialized to
L<Mojolicious::Renderer/"encoding"> if unset.

=head1 METHODS

L<Template::Provider::Mojo> inherits all methods from L<Template::Provider> and
implements the following new ones.

=head2 fetch

Returns a compiled template for the name specified. See L<Template::Provider/"fetch($name)">
for usage details.

=head2 load

Loads a template without parsing or compiling it. This is used by the
L<INSERT|Template::Manual::Directives/"INSERT"> directive.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Template>, L<Mojolicious::Renderer>, L<Mojolicious::Plugin::TemplateToolkit>


lib/Mojolicious/Plugin/TemplateToolkit.pm
package Mojolicious::Plugin::TemplateToolkit;
use Mojo::Base 'Mojolicious::Plugin';

use Carp;
use Mojo::Util qw(encode md5_sum);
use Template;
use Template::Provider::Mojo;

our $VERSION = '0.007';

sub register {
	my ($self, $app, $conf) = @_;
	
	my $tt_config = $conf->{template} || {};
	$tt_config->{MOJO_RENDERER} = $app->renderer;
	push @{$tt_config->{LOAD_TEMPLATES}}, Template::Provider::Mojo->new($tt_config);
	my $tt = Template->new($tt_config);
	
	$app->renderer->add_handler($conf->{name} || 'tt2' => sub {
		my ($renderer, $c, $output, $options) = @_;
		
		my $inline = $options->{inline};
		my $name = defined $inline ? md5_sum encode('UTF-8', $inline) : undef;
		return unless defined($name //= $renderer->template_name($options));
		
		my %params;
		
		# Helpers
		foreach my $method (grep { m/^\w+\z/ } keys %{$renderer->helpers}) {
			my $sub = $renderer->helpers->{$method};
			$params{$method} = sub { carp "Calling helpers directly in templates is deprecated. Use c.$method"; $c->$sub(@_) };
		}
		
		# Stash values
		$params{$_} = $c->stash->{$_} for grep { m/^\w+\z/ } keys %{$c->stash};
		$params{self} = $params{c} = $c;
		$params{h} = $c->helpers;
		
		# Inline
		if (defined $inline) {
			$c->app->log->debug(qq{Rendering inline template "$name"});
			$tt->process(\$inline, \%params, $output) or die $tt->error, "\n";
		}
		
		# File
		else {
			# Try template
			if (defined(my $path = $renderer->template_path($options))) {
				$c->app->log->debug(qq{Rendering template "$name"});
				$tt->process($name, \%params, $output) or die $tt->error, "\n";
			}
			
			# Try DATA section
			elsif (defined(my $d = $renderer->get_data_template($options))) {
				$c->app->log->debug(qq{Rendering template "$name" from DATA section});
				$tt->process(\$d, \%params, $output) or die $tt->error, "\n";
			}
			
			# No template
			else { $c->app->log->debug(qq{Template "$name" not found}) }
		}
	});
}

1;

=head1 NAME

Mojolicious::Plugin::TemplateToolkit - Template Toolkit renderer plugin for
Mojolicious

=head1 SYNOPSIS

 # Mojolicious
 $app->plugin('TemplateToolkit');
 $app->plugin(TemplateToolkit => {name => 'foo'});
 $app->plugin(TemplateToolkit => {template => {INTERPOLATE => 1}});
 
 # Mojolicious::Lite
 plugin 'TemplateToolkit';
 plugin TemplateToolkit => {name => 'foo'};
 plugin TemplateToolkit => {template => {INTERPOLATE => 1}});
 
 # Set as default handler
 $app->renderer->default_handler('tt2');
 
 # Render without setting as default handler
 $c->render(template => 'bar', handler => 'tt2');

=head1 DESCRIPTION

L<Mojolicious::Plugin::TemplateToolkit> is a renderer for C<tt2> or
C<Template Toolkit> templates. See L<Template> and L<Template::Manual> for
details on the C<Template Toolkit> format, and L<Mojolicious::Guides::Rendering>
for general information on rendering in L<Mojolicious>.

Along with template files, inline and data section templates can be rendered in
the standard Mojolicious fashion. Template files and data sections will be
retrieved using L<Mojolicious::Renderer> via L<Template::Provider::Mojo> for
both direct rendering and directives such as C<INCLUDE>. This means that
instead of specifying L<INCLUDE_PATH|Template::Manual::Config/"INCLUDE_PATH">,
you should set L<Mojolicious::Renderer/"paths"> to the appropriate paths.

 $app->renderer->paths(['/path/to/templates']);
 push @{$app->renderer->paths}, '/path/to/more/templates';

L<Mojolicious> stash values will be exposed directly as
L<variables|Template::Manual::Variables> in the templates, and the current
controller object will be available as C<c> or C<self>, similar to
L<Mojolicious::Plugin::EPRenderer>.

Helper methods can be called on the controller object as normal, as well as on
a L<proxy object|Mojolicious::Controller/"helpers"> available as C<h>. The
proxy object was previously needed for efficiency as the controller object used
C<AUTOLOAD> to call helper methods, but since Mojolicious 8.04 the controller
object uses a more efficient mechanism and the proxy object is no longer
needed. See L<Mojolicious::Plugin::DefaultHelpers> and
L<Mojolicious::Plugin::TagHelpers> for a list of all built-in helper methods.

Accessing helper methods directly as variables, rather than via the controller
or proxy object, is deprecated and may be removed in a future release.

 $c->stash(description => 'template engine');
 $c->stash(engines => [qw(Template::Toolkit Text::Template)]);
 
 [% FOREACH engine IN engines %]
   [% engine %] is a [% description %].
 [% END %]
 
 [% c.link_to('Template Toolkit', 'http://www.template-toolkit.org') %]
 
 [% c.param('foo') %]


=head1 OPTIONS

L<Mojolicious::Plugin::TemplateToolkit> supports the following options.

=head2 name

 # Mojolicious::Lite
 plugin TemplateToolkit => {name => 'foo'};

Handler name, defaults to C<tt2>.

=head2 template

 # Mojolicious::Lite
 plugin TemplateToolkit => {template => {INTERPOLATE => 1}};

Configuration values passed to L<Template> object used to render templates.
Note that L<Template::Provider::Mojo> will use L<Mojolicious::Renderer/"paths">
to find templates, not L<INCLUDE_PATH|Template::Manual::Config/"INCLUDE_PATH">
specified here.

=head1 METHODS

L<Mojolicious::Plugin::TemplateToolkit> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

 $plugin->register(Mojolicious->new);
 $plugin->register(Mojolicious->new, {name => 'foo'});

Register renderer in L<Mojolicious> application.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojolicious::Renderer>, L<Template>, L<Template::Provider::Mojo>
